### PR TITLE
Add tests to cover ISymbolExtensions.IsPrimaryConstructor

### LINKED ISSUE #8263: Add tests to cover ISymbolExtensions.IsPrimaryConstructor
Due to insufficient time, with this #8251 we are merging a method without proper coverage.

### PR DESCRIPTION
**Pull request by cristian-ambrosini-sonarsource**
Fixes #8263

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.UnitTest/Extensions/ISymbolExtensionsTest.cs [hunk 1]
```diff
@@ -148,6 +150,56 @@ End Sub
...
+    [DataRow("readonly struct MyReadonlyStruct { public MyReadonlyStruct() { } };", 0)]
+    public void IsPrimaryConstructor_CS(string code, int primaryConstructors)
+    {
+        var compilation = SolutionBuilder.Create()
+                                         .AddProject(AnalyzerLanguage.CSharp, createExtraEmptyFile: false)
+                                         .AddSnippet(code)
+                                         .GetCompilation(ParseOptionsHelper.CSharpLatest.First());
+
+        var semanticModel = compilation.GetSemanticModel(compilation.SyntaxTrees.First());
+        var typeSymbol = (INamedTypeSymbol)semanticModel.GetDeclaredSymbol(compilation.SyntaxTrees.First().GetRoot().DescendantNodes().First());
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Extensions/ISymbolExtensionsTest.cs:198]: Use `CreateSymbol` instead.

File: analyzers/tests/SonarAnalyzer.UnitTest/Extensions/ISymbolExtensionsTest.cs [hunk 2]
```diff
@@ -148,6 +150,56 @@ End Sub
         ISymbolExtensionsCommon.IsAutoProperty(CreateSymbol(code, AnalyzerLanguage.VisualBasic)).Should().BeFalse();
     }
 
+    [DataTestMethod]
+    [DataRow("class MyClass();", 1)]
+    [DataRow("class MyClass() { }", 1)]
+    [DataRow("class MyClass(int a) { }", 1)]
+    [DataRow("class MyClass { }", 0)]
+    [DataRow("class MyClass(int a) { MyClass() : this(1) { } };", 1)]
+    [DataRow("class MyClass { MyClass() { } };", 0)]
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Extensions/ISymbolExtensionsTest.cs:159]:
```suggestion
    [DataRow("class MyClass { MyClass() { } };", 0)]
    [DataRow("class Base(int i); class SymbolMember() : Base(1);", true)]
    [DataRow("""
    class Base(int i);
    class SymbolMember : Base
    {
        SymbolMember() : base(1) { }
    }
    """, false)]
```

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]

File: analyzers/tests/SonarAnalyzer.UnitTest/Extensions/ISymbolExtensionsTest.cs [hunk 3]
```diff
@@ -145,13 +147,68 @@ public void IsAutoProperty_NonpropertySymbol_VB()
...
+
+        methodSymbols.Count(ISymbolExtensionsCS.IsPrimaryConstructor).Should().Be(hasPrimaryConstructor ? 1 : 0);
+    }
+
+    private static ISymbol CreateSymbol(string snippet, bool firstOccurence, AnalyzerLanguage language, ParseOptions parseOptions = null)
     {
-        var (tree, semanticModel) = TestHelper.Compile(snippet, false, language);
-        var node = tree.GetRoot().DescendantNodes().Last(x => x.ToString().Contains(" SymbolMember"));
+        var (tree, semanticModel) = TestHelper.Compile(snippet, false, language, parseOptions: parseOptions);
+        var node = firstOccurence
```
Comment (cristian-ambrosini-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Extensions/ISymbolExtensionsTest.cs:209]: This is a very ugly implementation, but I was not able to come up with something more elegant for this problem.

Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/Extensions/ISymbolExtensionsTest.cs:209]: You can prepend unwanted names with @ instead:
`[DataRow("readonly struct @SymbolMember(int a) { public SymbolMember() : this(1) { } };")]`

File: analyzers/tests/SonarAnalyzer.UnitTest/Extensions/ISymbolExtensionsTest.cs [hunk 4]
```diff
@@ -145,13 +147,68 @@ public void IsAutoProperty_NonpropertySymbol_VB()
...
+    [DataRow("record class SymbolMember { SymbolMember() { } };", false)]
+    [DataRow("readonly struct SymbolMember();", true)]
+    [DataRow("readonly struct SymbolMember() { }", true)]
+    [DataRow("readonly struct SymbolMember(int a) { }", true)]
+    [DataRow("readonly struct SymbolMember { }", false)]
+    [DataRow("readonly struct SymbolMember(int a) { public SymbolMember() : this(1) { } };", true)]
+    [DataRow("readonly struct SymbolMember { public SymbolMember() { } };", false)]
+    public void IsPrimaryConstructor_CS(string code, bool hasPrimaryConstructor)
+    {
+        var typeSymbol = (INamedTypeSymbol)CreateSymbol(code, true, AnalyzerLanguage.CSharp, ParseOptionsHelper.CSharpLatest.First());
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Extensions/ISymbolExtensionsTest.cs:200]: Consider using `new CSharpParseOptions(LanguageVersion.Latest);` instead.

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]

File: analyzers/tests/SonarAnalyzer.UnitTest/Extensions/ISymbolExtensionsTest.cs (hunk 4 above)
Comment (cristian-ambrosini-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/Extensions/ISymbolExtensionsTest.cs:200]: Very smart solution! Thanks

File: analyzers/tests/SonarAnalyzer.UnitTest/Extensions/ISymbolExtensionsTest.cs [hunk 5]
```diff
@@ -19,10 +19,13 @@
  */
 
 extern alias vbnet;
+extern alias csharp;
 
+using Microsoft.CodeAnalysis.CSharp;
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Extensions/ISymbolExtensionsTest.cs:24]:
```suggestion
```

**Review by martin-strecker-sonarsource — APPROVED** [review]
LGTM. There seems to ba an invalid using, though.

**Pull request merged**